### PR TITLE
Destination BigQuery: Accept Dataset ID field prefixed by Project ID

### LINKED ISSUE #1192: Parsing Dataset names in BigQuery
## Tell us about the problem you're trying to solve
When specifying a Dataset_id, the Google UI often includes the project_id in its naming like this: `project_id:dataset_id`

## Describe the solution you’d like
In the BigQuery destination code, we could accept both syntax:

* project_id:dataset_id
* dataset_id

So it's smoother for the end user whether project_id is specified or not?

## Additional context

Example setup from a user in slack:
https://airbytehq.slack.com/archives/C019WJFR1JT/p1607060567071700?thread_ts=1607060325.069200&cid=C019WJFR1JT

image

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368134841734) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by koji-m**
## What
Closes: https://github.com/airbytehq/airbyte/issues/1192

## How
In the BigQuery destination connector's Dataset ID field, we could accept both syntax:

- project-id:dataset_id
- dataset_id

Make it error, if project-id in the Dataset ID field doesn't match the value in the Project ID field.

## Recommended reading order
1. `BigQueryDestinationTest.java`
2. `BigQueryDestination.java`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>
スクリーンショット 2021-12-01 23 12 40

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/8383)

### CODE REVIEW

**Comment by mkhokh-33** [comment]
Hi, @koji-m could you please update the branch and resolve conflicts on PR, tnx

**Comment by koji-m** [comment]
Hi, @mkhokh-33 
I've updated the branch and resolved conflicts. And integration-test passed at local.
Please review.

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public BigQueryDestination() {
   @Override
   public AirbyteConnectionStatus check(final JsonNode config) {
     try {
-      final String datasetId = config.get(BigQueryConsts.CONFIG_DATASET_ID).asText();
+      final String datasetId = BigQueryUtils.getDatasetId(config);
```
Comment (mkhokh-33) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:59]: BigQueryUtils.getDatasetId(config) used in 2 cases

1. BigQueryDestination.check()
2. BigQueryDestination.getConsumer() invoke inside BigQueryUploaderFactory.getUploader then BigQueryUtils.getSchema then BigQueryUtils.getDatasetId(config)

Could you pls consider to add integration test to BigQueryDestinationTest  to check that:
1. we can create connection with provided dataset
2. we can get consumer and write some data

**Review by mkhokh-33 — CHANGES_REQUESTED** [review]

**Comment by mkhokh-33** [comment]
> Hi, @mkhokh-33 I've updated the branch and resolved conflicts. And integration-test passed at local. Please review.

hi @koji-m,  thanks for your PR, I left some comments

**Comment by koji-m** [comment]
@mkhokh-33 Thank you for your review. I've fixed based on your review.

**Comment by mkhokh-33** [comment]
LGTM,
connectors base fails is known issue will be fixed by - https://github.com/airbytehq/airbyte/pull/9274
@sherifnada could you have a look please, integration test are passed - https://github.com/airbytehq/airbyte/pull/9282

**Comment by sherifnada** [comment]
Deferring the review to @edgao -- Edward, would you be able to review this contribution? 

Also, @mkhokh-33 if LGTM can you leave an approval?

**Review by mkhokh-33 — APPROVED** [review]

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryUtils.java [hunk 2]
```diff
@@ -162,6 +166,27 @@ public static JsonNode getGcsAvroJsonNodeConfig(final JsonNode config) {
     return gcsJsonNode;
   }
 
+  public static String getDatasetId(final JsonNode config) {
+    String datasetId = config.get(BigQueryConsts.CONFIG_DATASET_ID).asText();
+    Matcher matcher = DATASET_ID_PATTERN.matcher(datasetId);
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryUtils.java:171]: I think the regex makes it a little hard to follow the logic; string manipulation feels more clear:
```java
int colonIndex = datasetId.indexOf(":");
if (colonIndex != -1) {
  String projectId = datasetId.substring(0, colonIndex);
  // check equality against config.get(PROJECT_ID)
  // ...
}
// if colonIndex is -1, then this returns the entire string
// otherwise it returns everything after the colon
return datasetId.substring(colonIndex + 1);
```
This means we would lose the validation aspect (e.g. if someone enters `invalid*characters!` we'll accept it). But those errors would be detected when the connector runs its `check` method, so that feels OK to me

File: airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationTest.java [hunk 3]
```diff
@@ -369,4 +384,29 @@ private boolean isTablePartitioned(final BigQuery bigquery, final Dataset datase
...
+
+    public void accept(JsonNode config) {
+      consumer.accept(config);
+    }
+  }
+
+  private static Stream<Arguments> datasetIdResetterProvider() {
+    return Stream.of(
+      Arguments.arguments(new DatasetIdResetter(config -> {})),
+      Arguments.arguments(new DatasetIdResetter(
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationTest.java:402]: let's add a comment to explain what this is testing

**Review by edgao — CHANGES_REQUESTED** [review]
two small comments, but otherwise this looks good!

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryUtils.java (hunk 2 above)
Comment (koji-m) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryUtils.java:171]: I see, I agree with your suggestion. We don't need to check here what BigQuery checks.

**Comment by koji-m** [comment]
@edgao Thank you for your review. I've fixed based on your review.

**Review by edgao — APPROVED** [review]
nice! @mkhokh-33 can you take care of releasing this?

**Comment by edgao** [comment]
actually, lemme publish a separate PR first, will comment here once that finishes

**Comment by edgao** [comment]
alright @mkhokh-33 we're good to go here!

**Comment by mkhokh-33** [comment]
Hi @koji-m ,
could you please update docker images version in following files:
1.  https://github.com/koji-m/airbyte/blob/parsing-dataset-names-in-bq/airbyte-integrations/connectors/destination-bigquery/Dockerfile

LABEL io.airbyte.version=0.6.2

**we also need to change version for destination-bigquery-denormalized cause it use destination-bigquery implementation**
2. https://github.com/koji-m/airbyte/blob/parsing-dataset-names-in-bq/airbyte-integrations/connectors/destination-bigquery-denormalized/Dockerfile

 LABEL io.airbyte.version=0.2.3

3. https://github.com/koji-m/airbyte/blob/parsing-dataset-names-in-bq/airbyte-config/init/src/main/resources/seed/destination_definitions.yaml

- dockerImage: "airbyte/destination-bigquery:0.6.2"
- dockerImage: "airbyte/destination-bigquery-denormalized:0.2.3"

4. https://github.com/koji-m/airbyte/blob/parsing-dataset-names-in-bq/airbyte-config/init/src/main/resources/config/STANDARD_DESTINATION_DEFINITION/22f6c74f-5699-40ff-833c-4a879ea40133.json  "dockerImageTag": "0.6.2",

5. https://github.com/koji-m/airbyte/blob/parsing-dataset-names-in-bq/airbyte-config/init/src/main/resources/config/STANDARD_DESTINATION_DEFINITION/079d5540-f236-4294-ba7c-ade8fd918496.json

"dockerImageTag": "0.2.3",

6. https://github.com/koji-m/airbyte/blob/parsing-dataset-names-in-bq/airbyte-config/init/src/main/resources/seed/destination_specs.yaml
-dockerImage: "airbyte/destination-bigquery:0.6.2

- dockerImage: "airbyte/destination-bigquery-denormalized:0.2.3" 

Thus I could proceed with merge. Thanks

**Comment by koji-m** [comment]
@mkhokh-33 Sorry for the late response. I've updated the docker image version numbers. (with docs)

- destination-bigquery: 0.6.3 (master) -> 0.6.4 (this branch)
- destination-bigquery-denormalized: 0.2.3 (master) -> 0.2.4 (this branch)

**Comment by mkhokh-33** [comment]
> @mkhokh-33 Sorry for the late response. I've updated the docker image version numbers. (with docs)
> 
> * destination-bigquery: 0.6.3 (master) -> 0.6.4 (this branch)
> * destination-bigquery-denormalized: 0.2.3 (master) -> 0.2.4 (this branch)

@koji-m thank you, merged :+1:
